### PR TITLE
reset_column_information on update primary_key

### PR DESCRIPTION
**Pull request by bogdanvlviv**
Fix `ActiveRecord::AttributeMethods::PrimaryKey::ClassMethods::primary_key=`.
reset_column_information on update primary_key
r? @rafaelfranca

### CODE REVIEW

**Comment by rafaelfranca** [comment]
There are a lot of unrelated changes. Could you please focus on the change that show the problem?

File: activerecord/test/cases/batches_test.rb [hunk 1]
```diff
@@ -366,16 +442,40 @@ def test_in_batches_should_be_loaded
...
+    end
+
+    Zine.in_batches(of: 1, load: false) do |relation|
+      assert_not relation.loaded?
+    end
+  ensure
+    Zine.primary_key = 'id'
+  end
+
+  def test_in_batches_should_be_loaded_when_primary_key_is_title
```
Comment (bogdanvlviv) [inline activerecord/test/cases/batches_test.rb]: > There are a lot of unrelated changes. Could you please focus on the change that show the problem?

this is and test that have Zine.primary_key = 'title'. it dosn't work when run all tests in activerecord folder

Comment (rafaelfranca) [inline activerecord/test/cases/batches_test.rb]: It is also a order dependent problem. I think you need to all `reset_column_information`.

**Comment by bogdanvlviv** [comment]
Travis haven't applied!

**Comment by rafaelfranca** [comment]
OK, please investigate. The code is correct like I and Xavier are
explaining. The prof of this is if you run the test you added alone it will
pass, so it is something leaking state in the test suite. Unfortunately I
don’t have time to investigate, so if you want this PR to be merged please
do investigate.

**Comment by bogdanvlviv** [comment]
When to use method `reset_column_information` it works. Why are those [changes](https://github.com/rails/rails/pull/25821/files#diff-f0b4abc491880e438bf01f3f795e6ec1R214) in #25821 bad. This changes fixed this bug and i think it will work faster.


Also i have got this bug when run it in 
`$ rails console`
```ruby

Post.primary_key # => 'id'
Post.primary_key = 'title'
Post.in_batches(of: 1, load: true) do |relation|
  # uninterrupted cycle
end
```

**Comment by bogdanvlviv** [comment]
OK, i will investigate more.

**Comment by fxn** [comment]
@bogdanvlviv Let's take a step back. Could you please produce a minimal self-contained script to reproduce using the [Active Record bug report template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb)?

**Comment by bogdanvlviv** [comment]
I have decision. Please wait.

**Comment by bogdanvlviv** [comment]
Script to reproduce the bug:

```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  gem 'rails', '~> 5.0'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_in_batches_should_be_loaded_when_primary_key_is_title
    posts_scount = 100

    posts_scount.times { |i| Post.create!(title: "title#{i}") }

    assert_equal posts_scount, Post.count

    Post.in_batches(of: 1, load: true) do |relation|
      assert relation.loaded?
    end

    Post.primary_key = 'title'
    Post.in_batches(of: 1, load: true) do |relation|
      # uninterrupted cycle
      puts 'uninterrupted cycle :)'

      assert relation.loaded?
    end
  ensure
    Post.primary_key = 'id'
  end
end
```

**Comment by bogdanvlviv** [comment]
I found this solution [here](https://github.com/rails/rails/blob/0f91e2bc2591340369c7644678eab6c1d93043c1/activerecord/lib/active_record/model_schema.rb#L144). @rafaelfranca, @fxn, thanks for the help.

**Comment by rafaelfranca** [comment]
Thanks! r? @fxn said he will review

**Comment by fxn** [comment]
When you use the macro

```ruby
class Post < ActiveRecord::Base
  self.primary_key = 'title'
end
```

all works as usual:

```ruby
Post.create.id # => nil, there is no title
```

But if we remove the macro and AR learns about the PK once, then it does not expect a change in the PK:

```ruby
Post.create
Post.primary_key = 'title'
p Post.first.id # => 1
```

and this is because the PK of a table is not expected to change on the fly and [it is cached](https://github.com/rails/rails/blob/cf5bc0dd5359e61dcb93e719279c7198e68a0eb8/activerecord/lib/active_record/connection_adapters/schema_cache.rb#L24-L26).

So, as @rafaelfranca and I said, the implementation of batches is correct. The mismatch here is that AR *globally* does not support changing the PK arbitrarily at any point in time.

To me that makes sense a priori because PKs do not change and AR classes map to existing, fixed tables (though I don't have the history of these decisions). I only wonder if AR should err when reassigning or can do better somehow.

/cc @tenderlove @sgrif

**Comment by tenderlove** [comment]
Agree with @fxn.  PKs should definitely not be changing at runtime.  I think there is a method to reset the PK for a class, but I don't think it clears every possible cache.  I don't think we should assume that a class's PK will be changing at runtime though.

**Comment by bogdanvlviv** [comment]
@fxn i found pull request where added 'Reset column information' into method `ActiveRecord::ModelSchema::ClassMethods::table_name=` #5267

@jonleighton, could you help us to make the right decision?

**Comment by bogdanvlviv** [comment]
This PR isn't actual anymore.
